src/store/index-level.ts

```diff
@@ -1,213 +1,284 @@
-import type { Filter, RangeFilter } from '../types/message-types.js';
+import type { FilteredQuery } from '../types/event-log.js';
+import type { LevelWrapper } from './level-wrapper.js';
+import type { EqualFilter, Filter, OneOfFilter, RangeFilter } from '../types/message-types.js';
 import type { LevelWrapperBatchOperation, LevelWrapperIteratorOptions } from './level-wrapper.js';
 
-import { executeUnlessAborted } from '../index.js';
 import { flatten } from '../utils/object.js';
-import { createLevelDatabase, LevelWrapper } from './level-wrapper.js';
+import { SortOrder } from '../types/message-types.js';
+
+
+type IndexedItem<T> = {
+  value: T;
+  key: string;
+};
+
+const INDEX_SUBLEVEL_NAME = 'index';
 
 export interface IndexLevelOptions {
   signal?: AbortSignal;
 }
 
 /**
- * A LevelDB implementation for indexing the messages stored in the DWN.
+ * A LevelDB implementation for indexing the messages and events stored in the DWN.
  */
-export class IndexLevel {
-  config: IndexLevelConfig;
-
-  db: LevelWrapper<string>;
-
-  constructor(config: IndexLevelConfig) {
-    this.config = {
-      createLevelDatabase,
-      ...config
-    };
-
-    this.db = new LevelWrapper<string>({ ...this.config, valueEncoding: 'utf8' });
-  }
-
-  async open(): Promise<void> {
-    return this.db.open();
-  }
-
-  async close(): Promise<void> {
-    return this.db.close();
-  }
+export class IndexLevel<T> {
+  constructor(private db: LevelWrapper<string>) {}
 
   /**
-   * Adds indexes for a specific data/object/content.
-   * @param dataId ID of the data/object/content being indexed.
+   * Put an item into the index using information that will allow it to be queried for.
+   *
+   * @param tenant
+   * @param key a unique key that represents the item being indexed, this is also used as the cursor value in a query.
+   * @param value the value representing the data being indexed.
+   * @param indexes - (key-value pairs) to be included as part of indexing this item. Must include at least one indexing property.
+   * @param sortIndexes - (key-value pairs) to be used for sorting the index. Must include at least one sorting property.
+   * @param options IndexLevelOptions that include an AbortSignal.
    */
   async put(
     tenant: string,
-    dataId: string,
-    indexes: { [property: string]: unknown },
+    key: string,
+    value: T,
+    indexes: { [key:string]: unknown },
+    sortIndexes: { [key:string]: unknown },
     options?: IndexLevelOptions
   ): Promise<void> {
-    const partition = await executeUnlessAborted(this.db.partition(tenant), options?.signal);
+    // ensure sorted indexes are flat and exist
+    sortIndexes = flatten(sortIndexes);
+    if (!sortIndexes || Object.keys(sortIndexes).length === 0) {
+      throw new Error('must include at least one sorted index');
```

Comment:
Why must we include at least one sort index? If I understand correctly, since we didn't have sort indexes before, we effectively had zero sort indexes.

Reply:
This is part of generalizing IndexLevel to allow different sorting properties. Before this change all of the sorting was done after retrieving all of the records from the query() method.

Since we want to allow an efficient query for a cursor/watermark (more specifically the watermark for EventsQuery in this PR) we have to include a sorting property, that way we can have a specific key as a starting point in level, and know that everything coming after that key is sorted accordingly.